src/Lib3.hs

```diff
@@ -193,6 +194,138 @@ executeSql sql = case parseStatement sql of
+        updatedRows = map updateRow' rows
+    in Right updatedRows
+
+updateOriginalDataFrame :: DataFrame -> [(Row, Row)] -> DataFrame
+updateOriginalDataFrame (DataFrame columns oldRows) updatedRowTuples =
+    let updatedRows = map snd updatedRowTuples
+        unchangedRows = filter (\oldRow -> notElem oldRow (map fst updatedRowTuples)) oldRows
+        newRows = unchangedRows ++ updatedRows
+    in DataFrame columns newRows
+
+updateWithoutWhere :: (TableName, DataFrame) -> [(ColumnName, Value)] -> Execution (Either ErrorMessage DataFrame)
+updateWithoutWhere (tableName, dataFrame) updates = do
+    case updateAllRows dataFrame updates of
+        Left errMsg -> return $ Left errMsg
+        Right updatedDataFrame -> do
+            saveTable (tableName, updatedDataFrame)
+            return $ Right updatedDataFrame
+
+updateAllRows :: DataFrame -> [(ColumnName, Value)] -> Either ErrorMessage DataFrame
+updateAllRows (DataFrame columns rows) updates =
+    let updatedRows = map (\row -> updateRow row updates columns) rows
+    in Right (DataFrame columns updatedRows)
+
+updateRow :: Row -> [(ColumnName, Value)] -> [Column]-> Row
+updateRow row updates columns =
+    foldl (\acc (colName, value) -> updateValue colName value acc columns) row updates
+
+updateValue :: ColumnName -> Value -> Row -> [Column] -> Row
+updateValue colName newValue row columns =
+    case findColumnIndex colName columns of
+        Left _ -> row -- Column not found, do nothing
+        Right colIndex -> replaceAtIndex colIndex newValue row
+
+replaceAtIndex :: Int -> a -> [a] -> [a]
+replaceAtIndex _ _ [] = []
+replaceAtIndex i newVal (x:xs)
+    | i == 0    = newVal : xs
+    | otherwise = x : replaceAtIndex (i - 1) newVal xs
+
+-- insert functions
+
+insertWithColumns :: (TableName, DataFrame) -> Maybe [ColumnName] -> [Value] -> Execution (Either ErrorMessage DataFrame)
+insertWithColumns (tableName, loadedDataFrame) justColumns values = do
+    case insertByColumnsIntoDataFrame loadedDataFrame justColumns values of
+        Left errMsg -> return $ Left errMsg
+        Right updatedDataFrame -> do
+            saveTable (tableName, updatedDataFrame)
+            return $ Right updatedDataFrame
+
+insertByColumnsIntoDataFrame :: DataFrame -> Maybe [ColumnName] -> [Value] -> Either ErrorMessage DataFrame
+insertByColumnsIntoDataFrame (DataFrame columns rows) maybeColumns newValues =
+    let changedColumns = fromMaybe [] maybeColumns
+        newRow = map (\col -> case elemIndex col changedColumns of
+                                  Just index -> newValues !! index
+                                  Nothing -> NullValue) [col | Column col _ <- columns]
+    in Right $ DataFrame columns (rows ++ [newRow])
+
+insertWithoutColumns :: (TableName, DataFrame) -> [Value] -> Execution (Either ErrorMessage DataFrame)
+insertWithoutColumns (tableName, loadedDataFrame) values = do
+    validatedValues <- validateRowLength loadedDataFrame values
```

Comment:
I think this checks only that the number of values is correct, but does not check the type is correct? Seems like kind of similar functionality like in Lab1 3rd function (called by `parseTable` and `serializeTable` as well)

Reply:
maybe we could just append the row and reuse the function? just a thought

Reply:
Oh yeah totally, since we worked so much with Lib2, I didn't really check for useful functions in Lib1

Reply:
On second thought i might just leave it as is, because I couldn't find a way to match the return types. I would need to make another function just to make it work. It's possible I missed some easy approach though